src/basic/pidref.c

```diff
@@ -457,12 +457,36 @@ bool pidref_is_automatic(const PidRef *pidref) {
         return pidref && pid_is_automatic(pidref->pid);
 }
 
-static void pidref_hash_func(const PidRef *pidref, struct siphash *state) {
+void pidref_hash_func(const PidRef *pidref, struct siphash *state) {
         siphash24_compress_typesafe(pidref->pid, state);
 }
 
-static int pidref_compare_func(const PidRef *a, const PidRef *b) {
-        return CMP(a->pid, b->pid);
+int pidref_compare_func(const PidRef *a, const PidRef *b) {
+        int r;
+
+        assert(a);
+        assert(b);
+
+        r = CMP(pidref_is_set(a), pidref_is_set(b));
+        if (r != 0)
+                return r;
+
+        r = CMP(pidref_is_remote(a), pidref_is_remote(b));
+        if (r != 0)
+                return r;
+
+        r = CMP(pidref_is_automatic(a), pidref_is_automatic(b));
```

Comment:
Hmm, I think automatic should be checked before remote. They're effectively "not set" after all